test/utils/index.ts

```diff
@@ -13,4 +15,309 @@ describe('utils', () => {
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>other_chars\t other_chars\t</div>');
+		});
+
+		it('hex number', () => {
+			const entities = [
+				'&#x9;',
+				'&#x21;',
+				'&#x22;',
+				'&#x23;',
+				'&#x24;',
+				'&#x25;'
+			];
+			for (const e of entities) {
+				const value = decode_character_references(e);
+				const codePoint = parseInt(e.match(/\d+;?/)[0], 16);
+				assert.equal(value, String.fromCodePoint(codePoint));
+			}
+		});
+
+		it('hex number without semicolon', () => {
+			const list = [
+				'&#x9',
+				'&#x21',
+				'&#x22',
+				'&#x23',
+				'&#x24',
+				'&#x25'
+			];
+			for (const e of list) {
+				const value = decode_character_references(e);
+				const codePoint = parseInt(e.match(/\d+;?/)[0], 16);
+				assert.equal(value, String.fromCodePoint(codePoint));
+			}
+		});
+
+		it('hex number with other character 1', () => {
+			const input = '<div>&#x9</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>\t</div>');
+		});
+
+		it('hex number with other character 2', () => {
+			const input = '<div>&#x9 other_chars</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>\t other_chars</div>');
+		});
+
+		it('hex number with other character 3', () => {
+			// doesn't decode if illegal
+			const input = '<div>&a#x9</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, input);
+		});
+
+		it('hex number with other character 4', () => {
+			// doesn't decode if without word boundary
+			const input = '<div>&#x9hello</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, input);
+		});
+
+		it('hex number with other character 5', () => {
+			const input = '<div>&#x9 other_chars</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>\t other_chars</div>');
+		});
+
+		it('hex number with other character 6', () => {
+			const input = '<div>&#x9; other_chars</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>\t other_chars</div>');
+		});
+
+		it('hex number with other character 7', () => {
+			const input = '<div>other_chars&#x9; other_chars</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>other_chars\t other_chars</div>');
+		});
+
+		it('hex number with other character 8', () => {
+			const input = '<div>other_chars&#x9; other_chars&#x9;</div>';
+			const value = decode_character_references(input);
+			assert.equal(value, '<div>other_chars\t other_chars\t</div>');
+		});
+
+		it('escape character', () => {
+			const list = [
+				'&amp;',
+				'&copy;',
+				'&lt;',
+				'&middot;',
+				'&minusd;',
+				'&AMP;'
+			];
+			for (const e of list) {
+				const value = decode_character_references(e);
+				const char = e.match(/\w+;?/)[0];
+				const codePoint = entities[char];
+				assert.equal(1, 1);
```

Comment:
I think we don't need this.

Reply:
Sorry, This is my carelessness.I will delele this in next commit.